text/0000-async-stream.md

```diff
@@ -0,0 +1,900 @@
+
+Individual streams may choose to resume iteration, and so calling [`next`]
+again may or may not eventually yield `Some(Item)` again at some point.
+
+### Usage
+
+A user could create a stream as follows (Example taken from @yoshuawuyt's [implementation pull request](https://github.com/rust-lang/rust/pull/79023)).
+
+Creating a stream involves two steps: creating a `struct` to
+ hold the stream's state, and then implementing `Stream` for that
+ `struct`.
+
+ Let's make a stream named `Counter` which counts from `1` to `5`:
+
+```rust
+#![feature(async_stream)]
+# use core::stream::Stream;
+# use core::task::{Context, Poll};
+# use core::pin::Pin;
+
+// First, the struct:
+
+/// A stream which counts from one to five
+struct Counter {
+    count: usize,
+}
+
+// we want our count to start at one, so let's add a new() method to help.
+// This isn't strictly necessary, but is convenient. Note that we start
+// `count` at zero, we'll see why in `poll_next()`'s implementation below.
+impl Counter {
+    fn new() -> Counter {
+        Counter { count: 0 }
+    }
+}
+
+// Then, we implement `Stream` for our `Counter`:
+
+impl Stream for Counter {
+    // we will be counting with usize
+    type Item = usize;
+
+    // poll_next() is the only required method
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Increment our count. This is why we started at zero.
+        self.count += 1;
+
+        // Check to see if we've finished counting or not.
+        if self.count < 6 {
+            Poll::Ready(Some(self.count))
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+```
+
+## next method
+
+We should also implement a next method, similar to [the implementation in the futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+In general, we have purposefully kept the core trait definition minimal. 
+There are a number of useful extension methods that are available, for example, 
+in the `futures-util` crate, but we have not included them because they involve 
+closure arguments, and we have not yet finalized the design of async closures.
+
+However, the core `poll_next` method is unergonomic; it does not let you iterate 
+over the items coming out of the stream. Therefore, we include a few minimal 
+convenience methods that are not dependent on any unstable features, such as `next`.
+
+As @yoshuawuyts states in their [pull request which adds `core::stream::Stream` to the standard library](https://github.com/rust-lang/rust/pull/79023):
+
+Unlike `Iterator`, `Stream` makes a distinction between the `poll_next`
+method which is used when implementing a `Stream`, and the `next` method
+which is used when consuming a stream. Consumers of `Stream` only need to
+consider `next`, which when called, returns a future which yields
+`Option<Item>`.
+
+The future returned by `next` will yield `Some(Item)` as long as there are
```

Comment:
Especially when combined with the paragraph below about `Future` guarantees, it's not obvious whether "the future" here actually means multiple futures or whether this is a specific guarantee that *the same* future returned by `next` can be polled again after yielding the first element to get the second, third and so on.